### PR TITLE
nixos/ntfy-sh: init module

### PR DESCRIPTION
**Pull request by Tom-Hubrecht**
###### Description of changes

Add a module to configure the ntfy.sh service.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 1]
```diff
@@ -162,6 +162,8 @@ Use `configure.packages` instead.
 
 - A new module was added for the Saleae Logic device family, providing the options `hardware.saleae-logic.enable` and `hardware.saleae-logic.package`.
 
+- There is a new module to allow configuring settings for the `ntfy.sh` service.
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2211.section.md:165]: It is in the wrong section, it should be in "New Services" and you should put links to the option

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+let
+  cfg = config.services.ntfy-sh;
+
+  settingsFormat = pkgs.formats.yaml { };
+in
+
+{
+  options.services.ntfy-sh = {
+    enable = mkEnableOption "[ntfy-sh](https://ntfy.sh), a push notification service";
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/ntfy-sh.nix:13]: I do not know if `mkEnableOption` pass this to `lib.mdDoc`, maybe @pennae do know better than me about this?

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    settings = mkOption {
+      type = types.submodule {
+
+        freeformType = settingsFormat.type;
+
+        options.base-url = mkOption {
+          type = types.str;
+          default = "";
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/ntfy-sh.nix:22]: Is this setting necessary? If so, why provide a default for it? Let it fail so the user can fix it

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+          description = "Public facing base URL of the service.";
+          example = "https://ntfy.sh";
+        };
+      };
+
+      default = { };
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/ntfy-sh.nix:28]: Can the software run with no setting by default?

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+      description = ''
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/ntfy-sh.nix:30]: You need `lib.mdDoc ''xxx''` in the front of your description

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      {
+        enable = true;
+
+        description = "Push notifications server";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+
+        serviceConfig = {
+          ExecStart = "${pkgs.ntfy-sh}/bin/ntfy serve -c ${configuration}";
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/ntfy-sh.nix:50]: You should add a `package` option and reference `cfg.package` here

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.ntfy-sh =
+      let
+        configuration = settingsFormat.generate "server.yml" cfg.settings;
+      in
+      {
+        enable = true;
```
Comment (RaitoBezarius) [inline nixos/modules/services/misc/ntfy-sh.nix:42]: This is not necessary, systemd services are enabled by default

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
It would be great to have a trivial NixOS test to test if the software run as expected and will help for the future maintenance.

**Comment by RaitoBezarius** [comment]
Also, please squash your commits into the first one.

File: nixos/modules/services/misc/ntfy-sh.nix (hunk 4 above)
Comment (Tom-Hubrecht) [reply nixos/modules/services/misc/ntfy-sh.nix:28]: Yes, if no settings are given, the server will simply listen on :80

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 1 above)
Comment (Tom-Hubrecht) [reply nixos/doc/manual/release-notes/rl-2211.section.md]: Moved

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        configuration = settingsFormat.generate "server.yml" cfg.settings;
+      in
+      {
+        description = "Push notifications server";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/ntfy serve -c ${configuration}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/ntfy-sh.nix:45]: Some hardening wouldn't hurt.

File: nixos/tests/ntfy-sh.nix [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    msg = "Test notification"
+
+    machine.wait_for_unit("multi-user.target")
+
+    machine.succeed(f"curl -d '{msg}' localhost:80/test")
+
+    notif = machine.succeed("curl -s localhost:80/test/json?poll=1 | jq '.message'")
+
+    if notif != f'"{msg}"\n':
+      raise Exception(f"Wrong message: '{notif}', expected '{msg}'")
```
Comment (RaitoBezarius) [inline nixos/tests/ntfy-sh.nix:19]: Use `assert notif == f'"{msg}"', 'Wrong message'`

File: nixos/tests/ntfy-sh.nix [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import ./make-test-python.nix {
```
Comment (RaitoBezarius) [inline nixos/tests/ntfy-sh.nix:1]: Please add this test in the test list (where `handleTests` are).

**Comment by RaitoBezarius** [comment]
@ofborg ntfy-sh

**Review by RaitoBezarius — APPROVED** [review]
Looks good to me.

**Comment by Tom-Hubrecht** [comment]
friendly ping @SuperSandro2000

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+          ProtectSystem = "full";
+          ProtectKernelTunables = "yes";
+          ProtectKernelModules = "yes";
+          ProtectKernelLogs = "yes";
+          ProtectControlGroups = "yes";
+          PrivateDevices = "yes";
+          RestrictSUIDSGID = "yes";
+          RestrictNamespaces = "yes";
+          RestrictRealtime = "yes";
+          MemoryDenyWriteExecute = "yes";
```
Comment (winterqt) [inline nixos/modules/services/misc/ntfy-sh.nix:59]: Please replace every instance of `"yes"` with `true` -- our systemd module converts `true` to `"yes"` automatically.

**Review by winterqt — CHANGES_REQUESTED** [review]

File: nixos/tests/ntfy-sh.nix [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  };
+
+  testScript = ''
+    msg = "Test notification"
+
+    machine.wait_for_unit("multi-user.target")
+
+    machine.succeed(f"curl -d '{msg}' localhost:80/test")
+
+    notif = machine.succeed("curl -s localhost:80/test/json?poll=1 | jq '.message'")
```
Comment (winterqt) [inline nixos/tests/ntfy-sh.nix:16]: Maybe this should be changed to actually parsing the output with Python's JSON module? Maybe not, idk, just a thought.

**Comment by winterqt** [comment]
(I requested @arjan-s's review since they maintain the package -- is there a reason you removed them?)

**Comment by arjan-s** [comment]
@winterqt that's just how github works, I experienced the same in my own PRs. When you click the "re-request review" button next to a reviewer's name, github automatically removes all other reviewers.

In any case, I'll review this shortly. I just didn't get around to it yet.

File: nixos/modules/services/misc/ntfy-sh.nix [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    };
+
+    settings = mkOption {
+      type = types.submodule { freeformType = settingsFormat.type; };
+
+      default = { };
+
+      description = mdDoc ''
+        Configuration for ntfy.sh, supported values are [here](https://ntfy.sh/docs/config/#config-options).
+      '';
```
Comment (arjan-s) [inline nixos/modules/services/misc/ntfy-sh.nix:29]:
```suggestion
      '';
      example = {
        listen-http = ":8080";
      };
```

**Review by arjan-s — CHANGES_REQUESTED** [review]
Just a minor addition. But looks good to me otherwise.
Tested it, and it works as expected.
Thanks!

**Comment by winterqt** [comment]
> @winterqt that's just how github works, I experienced the same in my own PRs. When you click the "re-request review" button next to a reviewer's name, github automatically removes all other reviewers.

That's strange, and smells like a bug. *sighs*

Sorry if it sounded like I was accusing you, @Tom-Hubrecht.

**Comment by Tom-Hubrecht** [comment]
> > @winterqt that's just how github works, I experienced the same in my own PRs. When you click the "re-request review" button next to a reviewer's name, github automatically removes all other reviewers.
> 
> That's strange, and smells like a bug. _sighs_ though knowing GitHub it could easily be intentional sleepy
> 
> Sorry if it sounded like I was accusing you, @Tom-Hubrecht.

No Problem !

File: nixos/modules/services/misc/ntfy-sh.nix (hunk 13 above)
Comment (Tom-Hubrecht) [reply nixos/modules/services/misc/ntfy-sh.nix:35]: Done ^^

Comment (arjan-s) [reply nixos/modules/services/misc/ntfy-sh.nix:35]: Ah, thanks! I looks a multiple examples and found both with and without that.

**Review by arjan-s — APPROVED** [review]
LGTM!

File: nixos/tests/ntfy-sh.nix (hunk 12 above)
Comment (winterqt) [reply nixos/tests/ntfy-sh.nix:16]: @Tom-Hubrecht @arjan-s Any thoughts on this? Last thing to discuss before it's good to go :)

(To be clear, I'm fine with either really, but would just appreciate feedback. Thanks!)

Comment (arjan-s) [reply nixos/tests/ntfy-sh.nix:16]: Ahh, I missed the fact that `testScript` is actual Python code. In that case, yes, I would parse it using the built-in JSON module.

Comment (Tom-Hubrecht) [reply nixos/tests/ntfy-sh.nix]: Done, it is better to parse the json in python indeed

**Comment by arjan-s** [comment]
Sorry for the possible extra work, but I think the service would be even better if it could be started as an unprivileged user. When this runs on ports above 1024, there's no reason to run it as root. Perhaps this could even be configurable?

**Comment by winterqt** [comment]
I somehow didn't notice this.

Yes, this is a must.

My suggestion is to use `DynamicUser = true` and adding the `CAP_NET_BIND_SERVICE` capability (by setting `AmbientCapabilities` and `CapabilityBoundingSet`) if the port is less than 1024.

**Comment by SuperSandro2000** [comment]
Amending that in a bit since I want to deploy ntfy.sh.

**Comment by arjan-s** [comment]
@SuperSandro2000 Nice, but why not use `DynamicUser`? Does it interfere with the CLI?

**Comment by winterqt** [comment]
It definitely wouldn't interfere with it, as the configuration file in `/etc` is world-readable.

**Comment by SuperSandro2000** [comment]
The cli also changes files in the state directory and that does not play nicely with DynamicUser.

**Pull request merged**